imageio/imageio-tiff/src/main/java/com/twelvemonkeys/imageio/plugins/tiff/CCITTFaxDecoderStream.java

```diff
@@ -152,39 +152,50 @@ public CCITTFaxDecoderStream(final InputStream stream, final int columns, final
     static int findCompressionType(final int type, final InputStream in) throws IOException {
         // Discover possible incorrect type, revert to RLE
         if (type == TIFFExtension.COMPRESSION_CCITT_T4 && in.markSupported()) {
-            byte[] streamData = new byte[32];
+            int limit = 500;
 
             try {
-                in.mark(streamData.length);
+                in.mark(limit);
 
-                int offset = 0;
-                while (offset < streamData.length) {
-                    int read = in.read(streamData, offset, streamData.length - offset);
-                    if (read <= 0) {
-                        break;
-                    }
-
-                    offset += read;
+                int first = in.read();
+                int second = in.read();
+                if (first == -1 || second == -1) {
+                    // stream to short
+                    return type;
                 }
-            }
-            finally {
-                in.reset();
-            }
+                else if (first == 0 && (((byte) second) >> 4 == 1 || ((byte) second) == 1)) {
+                    // correct, starts with EOL or byte aligned EOL
+                    return type;
+                }
+                short b = (short) (((((byte) first) << 8) + ((byte) second)) >> 4);
+                int limitBits = limit * 8;
+                int read = second;
+                byte streamByte = (byte) read;
+                for (int i = 12; i < limitBits; i++) {
+                    if (i % 8 == 0) {
+                        read = in.read();
```

Comment:
Using as many as 500 single-byte `read`s over one array `read` might impact performance negatively, but I guess it's not a deal breaker.

But would it be hard to read blocks of bytes instead?

Reply:
Fair enough! 😀